File: pkg/protocol/messages.go
package protocol

import (
	"encoding/json"
	"time"
)

// WebSocket timing constants.
const (
	// WSWriteWait is the time allowed to write a message.
	WSWriteWait = 30 * time.Second

	// WSPongWait is the time to wait for a pong response.
	WSPongWait = 15 * time.Second

	// WSPingPeriod is how often to send pings (must be < PongWait).
	WSPingPeriod = 5 * time.Second

	// WSMaxMessageSize is the maximum message size in bytes (50MB).
	WSMaxMessageSize = 50 * 1024 * 1024

	// WSChunkSize is the size for binary chunks (1MB).
	WSChunkSize = 1024 * 1024

	// WSRequestTimeout is the timeout for request/response operations.
	WSRequestTimeout = 30 * time.Second
)

// MessageType identifies the type of WebSocket message.
type MessageType string

const (
	// Connection management
	MsgTypeHubConnected MessageType = "hub_connected" // Hub → Agent: handshake
	MsgTypeAgentStatus  MessageType = "agent_status"  // Agent → Hub: handshake response

	// Authentication / Pairing
	MsgTypePairingRequired MessageType = "pairing_required" // Agent → Hub: requires pairing
	MsgTypePairConfirm     MessageType = "pair_confirm"     // Hub → Agent: confirm pairing code
	MsgTypePairSuccess     MessageType = "pair_success"     // Agent → Hub: pairing successful
	MsgTypePairFailed      MessageType = "pair_failed"      // Agent → Hub: pairing failed

	// Requests from Hub to Agent
	MsgTypePing           MessageType = "ping"
	MsgTypeGetInfo        MessageType = "get_info"
	MsgTypeGetConfig      MessageType = "get_config"
	MsgTypeGetSteamUsers  MessageType = "get_steam_users"
	MsgTypeListShortcuts  MessageType = "list_shortcuts"
	MsgTypeCreateShortcut MessageType = "create_shortcut"
	MsgTypeDeleteShortcut MessageType = "delete_shortcut"
	MsgTypeDeleteGame     MessageType = "delete_game" // Agent handles everything internally
	MsgTypeApplyArtwork      MessageType = "apply_artwork"
	MsgTypeSendArtworkImage  MessageType = "send_artwork_image"  // Hub → Agent: binary image data
	MsgTypeRestartSteam      MessageType = "restart_steam"
	MsgTypeInitUpload     MessageType = "init_upload"
	MsgTypeUploadChunk    MessageType = "upload_chunk"
	MsgTypeCompleteUpload MessageType = "complete_upload"
	MsgTypeCancelUpload   MessageType = "cancel_upload"

	// Responses from Agent to Hub
	MsgTypePong              MessageType = "pong"
	MsgTypeInfoResponse      MessageType = "info_response"
	MsgTypeConfigResponse    MessageType = "config_response"
	MsgTypeSteamUsersResponse MessageType = "steam_users_response"
	MsgTypeShortcutsResponse MessageType = "shortcuts_response"
	MsgTypeArtworkResponse        MessageType = "artwork_response"
	MsgTypeArtworkImageResponse   MessageType = "artwork_image_response" // Agent → Hub: ack for binary artwork
	MsgTypeSteamResponse          MessageType = "steam_response"
	MsgTypeUploadInitResponse MessageType = "upload_init_response"
	MsgTypeUploadChunkResponse MessageType = "upload_chunk_response"
	MsgTypeOperationResult   MessageType = "operation_result"
	MsgTypeError             MessageType = "error"

	// Events from Agent to Hub (push notifications)
	MsgTypeUploadProgress MessageType = "upload_progress"
	MsgTypeOperationEvent MessageType = "operation_event"
)

// WSError represents an error in a WebSocket message.
type WSError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

// Message is the envelope for all WebSocket communication.
type Message struct {
	ID      string          `json:"id"`
	Type    MessageType     `json:"type"`
	Payload json.RawMessage `json:"payload,omitempty"`
	Error   *WSError        `json:"error,omitempty"`
}

// NewMessage creates a new message with the given type and payload.
func NewMessage(id string, msgType MessageType, payload any) (*Message, error) {
	var raw json.RawMessage
	if payload != nil {
		data, err := json.Marshal(payload)
		if err != nil {
			return nil, err
		}
		raw = data
	}
	return &Message{ID: id, Type: msgType, Payload: raw}, nil
}

// ParsePayload unmarshals the payload into the given type.
func (m *Message) ParsePayload(v any) error {
	if m.Payload == nil {
		return nil
	}
	return json.Unmarshal(m.Payload, v)
}

// NewErrorMessage creates an error response message.
func NewErrorMessage(id string, code int, message string) *Message {
	return &Message{
		ID:   id,
		Type: MsgTypeError,
		Error: &WSError{
			Code:    code,
			Message: message,
		},
	}
}

// Reply creates a response message for this request.
func (m *Message) Reply(msgType MessageType, payload any) (*Message, error) {
	return NewMessage(m.ID, msgType, payload)
}

// ReplyError creates an error response for this request.
func (m *Message) ReplyError(code int, message string) *Message {
	return NewErrorMessage(m.ID, code, message)
}

// Common WebSocket error codes.
const (
	WSErrCodeBadRequest     = 400
	WSErrCodeUnauthorized   = 401
	WSErrCodeNotFound       = 404
	WSErrCodeNotAccepted    = 406
	WSErrCodeConflict       = 409
	WSErrCodeInternal       = 500
	WSErrCodeNotImplemented = 501
)

// Request payloads

// InitUploadRequest starts a new upload session.
type InitUploadRequest struct {
	Config     UploadConfig `json:"config"`
	TotalSize  int64        `json:"totalSize"`
	FileCount  int          `json:"fileCount"`
	ResumeFrom int64        `json:"resumeFrom,omitempty"`
}

// UploadChunkRequest sends a chunk of data.
type UploadChunkRequest struct {
	UploadID string `json:"uploadId"`
	Offset   int64  `json:"offset"`
	Data     []byte `json:"data"`
	FilePath string `json:"filePath"`
	IsLast   bool   `json:"isLast"`
}

// CompleteUploadRequest finalizes an upload.
type CompleteUploadRequest struct {
	UploadID       string `json:"uploadId"`
	CreateShortcut bool   `json:"createShortcut"`
}

// CancelUploadRequest cancels an active upload.
type CancelUploadRequest struct {
	UploadID string `json:"uploadId"`
}

// CreateShortcutRequest creates a Steam shortcut.
type CreateShortcutRequest struct {
	UserID   uint32         `json:"userId"`
	Shortcut ShortcutConfig `json:"shortcut"`
}

// DeleteShortcutRequest removes a Steam shortcut.
type DeleteShortcutRequest struct {
	UserID  uint32 `json:"userId"`
	AppID   uint32 `json:"appId,omitempty"`
	Name    string `json:"name,omitempty"`
}

// ListShortcutsRequest lists shortcuts for a user.
type ListShortcutsRequest struct {
	UserID uint32 `json:"userId"`
}

// Response payloads

// InfoResponse contains agent information.
type InfoResponse struct {
	Agent AgentInfo `json:"agent"`
}

// InitUploadResponse acknowledges upload initialization.
type InitUploadResponse struct {
	UploadID   string `json:"uploadId"`
	ResumeFrom int64  `json:"resumeFrom"`
}

// UploadChunkResponse acknowledges a chunk.
type UploadChunkResponse struct {
	UploadID    string `json:"uploadId"`
	BytesWritten int64 `json:"bytesWritten"`
	TotalWritten int64 `json:"totalWritten"`
}

// CompleteUploadResponse confirms upload completion.
type CompleteUploadResponse struct {
	UploadID string `json:"uploadId"`
	Success  bool   `json:"success"`
}

// ShortcutResponse contains shortcut operation result.
type ShortcutResponse struct {
	Success   bool           `json:"success"`
	Shortcuts []ShortcutInfo `json:"shortcuts,omitempty"`
}

// SteamStatusResponse contains Steam status.
type SteamStatusResponse struct {
	Running bool   `json:"running"`
	Path    string `json:"path,omitempty"`
}

// ErrorResponse contains error details.
type ErrorResponse struct {
	Code    string `json:"code"`
	Message string `json:"message"`
	Details string `json:"details,omitempty"`
}

// Connection payloads

// HubConnectedRequest is sent when a Hub connects to an Agent.
type HubConnectedRequest struct {
	Name     string `json:"name"`
	Version  string `json:"version"`
	Platform string `json:"platform,omitempty"` // Hub platform (windows, linux, darwin)
	HubID    string `json:"hubId,omitempty"`    // Unique Hub identifier
	Token    string `json:"token,omitempty"`    // Auth token from previous pairing
}

// AgentStatusResponse is the Agent's response to a Hub connection.
type AgentStatusResponse struct {
	Name              string `json:"name"`
	Version           string `json:"version"`
	Platform          string `json:"platform"`
	AcceptConnections bool   `json:"acceptConnections"`
}

// PairingRequiredResponse is sent when a Hub needs to pair.
type PairingRequiredResponse struct {
	Code      string `json:"code"`      // 6-digit pairing code
	ExpiresIn int    `json:"expiresIn"` // Seconds until expiration
}

// PairConfirmRequest is sent by Hub to confirm pairing.
type PairConfirmRequest struct {
	Code string `json:"code"` // 6-digit code entered by user
}

// PairSuccessResponse is sent when pairing is successful.
type PairSuccessResponse struct {
	Token string `json:"token"` // Auth token for future connections
}

// PairFailedResponse is sent when pairing fails.
type PairFailedResponse struct {
	Reason string `json:"reason"` // Failure reason
}

// Config payloads

// ConfigResponse contains agent configuration.
type ConfigResponse struct {
	InstallPath string `json:"installPath"`
}

// Steam payloads

// SteamUser represents a Steam user (matches steam.User).
type SteamUser struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	AvatarURL   string `json:"avatarUrl,omitempty"`
	LastLoginAt int64  `json:"lastLoginAt,omitempty"`
}

// SteamUsersResponse contains the list of Steam users.
type SteamUsersResponse struct {
	Users []SteamUser `json:"users"`
}

// ShortcutsListResponse contains the list of shortcuts.
type ShortcutsListResponse struct {
	Shortcuts []ShortcutInfo `json:"shortcuts"`
}

// CreateShortcutResponse contains the result of shortcut creation.
type CreateShortcutResponse struct {
	AppID          uint32 `json:"appId"`
	SteamRestarted bool   `json:"steamRestarted,omitempty"`
}

// DeleteShortcutRequest with restart option.
type DeleteShortcutWithRestartRequest struct {
	UserID       string `json:"userId"`
	AppID        uint32 `json:"appId"`
	RestartSteam bool   `json:"restartSteam,omitempty"`
}

// DeleteGameRequest requests deletion of a game. Agent handles everything internally.
type DeleteGameRequest struct {
	AppID uint32 `json:"appId"`
}

// DeleteGameResponse contains the result of game deletion.
type DeleteGameResponse struct {
	Status         string `json:"status"`
	GameName       string `json:"gameName"`
	SteamRestarted bool   `json:"steamRestarted"`
}

// Artwork payloads

// ApplyArtworkRequest requests artwork application.
type ApplyArtworkRequest struct {
	UserID  string         `json:"userId"`
	AppID   uint32         `json:"appId"`
	Artwork *ArtworkConfig `json:"artwork"`
}

// ArtworkResponse contains artwork operation result.
type ArtworkResponse struct {
	Applied []string        `json:"applied"`
	Failed  []ArtworkFailed `json:"failed,omitempty"`
}

// ArtworkFailed represents a failed artwork application.
type ArtworkFailed struct {
	Type  string `json:"type"`
	Error string `json:"error"`
}

// Operation payloads

// OperationResult is a generic result for operations.
type OperationResult struct {
	Success bool   `json:"success"`
	Message string `json:"message,omitempty"`
}

// OperationEvent is a push notification for operation progress.
type OperationEvent struct {
	Type     string  `json:"type"`     // "install", "delete"
	Status   string  `json:"status"`   // "start", "progress", "complete", "error"
	GameName string  `json:"gameName"`
	Progress float64 `json:"progress"` // 0-100
	Message  string  `json:"message,omitempty"`
}

// Upload payloads (extended)

// InitUploadRequestFull includes file entries for the upload.
type InitUploadRequestFull struct {
	Config    UploadConfig `json:"config"`
	TotalSize int64        `json:"totalSize"`
	Files     []FileEntry  `json:"files"`
}

// FileEntry represents a file in the upload manifest.
type FileEntry struct {
	RelativePath string `json:"relativePath"`
	Size         int64  `json:"size"`
}

// InitUploadResponseFull includes chunk size configuration.
type InitUploadResponseFull struct {
	UploadID   string           `json:"uploadId"`
	ChunkSize  int              `json:"chunkSize"`
	ResumeFrom map[string]int64 `json:"resumeFrom,omitempty"`
}

// UploadChunkRequestFull includes all chunk metadata.
type UploadChunkRequestFull struct {
	UploadID string `json:"uploadId"`
	FilePath string `json:"filePath"`
	Offset   int64  `json:"offset"`
	Size     int    `json:"size"`
	Checksum string `json:"checksum,omitempty"`
	// Data is sent as binary message, not in JSON
}

// CompleteUploadRequestFull includes shortcut configuration.
type CompleteUploadRequestFull struct {
	UploadID       string          `json:"uploadId"`
	CreateShortcut bool            `json:"createShortcut"`
	Shortcut       *ShortcutConfig `json:"shortcut,omitempty"`
}

// CompleteUploadResponseFull includes the result path and appID.
type CompleteUploadResponseFull struct {
	Success bool   `json:"success"`
	Path    string `json:"path,omitempty"`
	AppID   uint32 `json:"appId,omitempty"`
}

// UploadProgressEvent is sent during upload to report progress.
type UploadProgressEvent struct {
	UploadID         string  `json:"uploadId"`
	TransferredBytes int64   `json:"transferredBytes"`
	TotalBytes       int64   `json:"totalBytes"`
	CurrentFile      string  `json:"currentFile,omitempty"`
	Percentage       float64 `json:"percentage"`
}

// ArtworkImageResponse contains the result of a binary artwork image transfer.
type ArtworkImageResponse struct {
	Success     bool   `json:"success"`
	ArtworkType string `json:"artworkType"`
	Error       string `json:"error,omitempty"`
}

// Steam control payloads

// RestartSteamResponse contains the result of Steam restart.
type RestartSteamResponse struct {
	Success bool   `json:"success"`
	Message string `json:"message"`
}

File: crates/protocol/src/messages.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::types::{AgentInfo, ArtworkConfig, ShortcutConfig, ShortcutInfo, UploadConfig};

// ---------------------------------------------------------------------------
// Request payloads
// ---------------------------------------------------------------------------

/// Starts a new upload session.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InitUploadRequest {
    pub config: UploadConfig,
    pub total_size: i64,
    pub file_count: i32,
    #[serde(default, skip_serializing_if = "is_zero_i64")]
    pub resume_from: i64,
}

/// Sends a chunk of upload data.
///
/// The `data` field is base64-encoded in JSON to match Go's `[]byte` serialization.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UploadChunkRequest {
    pub upload_id: String,
    pub offset: i64,
    #[serde(with = "base64_bytes")]
    pub data: Vec<u8>,
    pub file_path: String,
    pub is_last: bool,
}

/// Finalizes an upload.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompleteUploadRequest {
    pub upload_id: String,
    pub create_shortcut: bool,
}

/// Cancels an active upload.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CancelUploadRequest {
    pub upload_id: String,
}

/// Sets which log levels the agent should collect.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SetConsoleLogFilterRequest {
    pub level_mask: u32,
}

/// Enables or disables console log streaming.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SetConsoleLogEnabledRequest {
    pub enabled: bool,
}

/// Creates a Steam shortcut.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateShortcutRequest {
    pub user_id: u32,
    pub shortcut: ShortcutConfig,
}

/// Removes a Steam shortcut.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteShortcutRequest {
    pub user_id: u32,
    #[serde(default, skip_serializing_if = "is_zero_u32")]
    pub app_id: u32,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub name: String,
}

/// Lists shortcuts for a user.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListShortcutsRequest {
    pub user_id: u32,
}

/// Requests artwork application.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApplyArtworkRequest {
    pub user_id: String,
    pub app_id: u32,
    pub artwork: Option<ArtworkConfig>,
}

/// Requests deletion of a game (agent handles everything internally).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteGameRequest {
    pub app_id: u32,
}

/// Enables or disables the game log wrapper for a specific game.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SetGameLogWrapperRequest {
    pub app_id: u32,
    pub enabled: bool,
}

// ---------------------------------------------------------------------------
// Connection payloads
// ---------------------------------------------------------------------------

/// Sent when a Hub connects to an Agent.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HubConnectedRequest {
    pub name: String,
    pub version: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub platform: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub hub_id: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub token: String,
    /// Protocol version advertised by the Hub (0 = legacy/pre-negotiation).
    #[serde(default, skip_serializing_if = "is_zero_u32")]
    pub protocol_version: u32,
}

/// Agent's response to a Hub connection.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentStatusResponse {
    pub name: String,
    pub version: String,
    pub platform: String,
    pub accept_connections: bool,
    pub telemetry_enabled: bool,
    pub telemetry_interval: i32,
    pub console_log_enabled: bool,
    /// Protocol version advertised by the Agent (0 = legacy/pre-negotiation).
    #[serde(default, skip_serializing_if = "is_zero_u32")]
    pub protocol_version: u32,
}

/// Sent when a Hub needs to pair.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PairingRequiredResponse {
    pub code: String,
    pub expires_in: i32,
}

/// Sent by Hub to confirm pairing.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PairConfirmRequest {
    pub code: String,
}

/// Sent when pairing is successful.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PairSuccessResponse {
    pub token: String,
}

/// Sent when pairing fails.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PairFailedResponse {
    pub reason: String,
}

// ---------------------------------------------------------------------------
// Response payloads
// ---------------------------------------------------------------------------

/// Confirms the applied log level mask.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SetConsoleLogFilterResponse {
    pub level_mask: u32,
}

/// Confirms the console log enabled state.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SetConsoleLogEnabledResponse {
    pub enabled: bool,
}

/// Contains agent information.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct InfoResponse {
    pub agent: AgentInfo,
}

/// Acknowledges upload initialization.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InitUploadResponse {
    pub upload_id: String,
    pub resume_from: i64,
}

/// Acknowledges a chunk.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UploadChunkResponse {
    pub upload_id: String,
    pub bytes_written: i64,
    pub total_written: i64,
}

/// Confirms upload completion.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompleteUploadResponse {
    pub upload_id: String,
    pub success: bool,
}

/// Contains shortcut operation result.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ShortcutResponse {
    pub success: bool,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub shortcuts: Vec<ShortcutInfo>,
}

/// Contains Steam status.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SteamStatusResponse {
    pub running: bool,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub path: String,
}

/// Contains error details.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ErrorResponse {
    pub code: String,
    pub message: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub details: String,
}

/// Agent configuration.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConfigResponse {
    pub install_path: String,
}

// ---------------------------------------------------------------------------
// Steam payloads
// ---------------------------------------------------------------------------

/// A Steam user (matches `steam.User` structure).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SteamUser {
    pub id: String,
    #[serde(default)]
    pub name: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub avatar_url: String,
    #[serde(default, skip_serializing_if = "is_zero_i64")]
    pub last_login_at: i64,
}

/// List of Steam users.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SteamUsersResponse {
    pub users: Vec<SteamUser>,
}

/// List of shortcuts.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ShortcutsListResponse {
    pub shortcuts: Vec<ShortcutInfo>,
}

/// Result of shortcut creation.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateShortcutResponse {
    pub app_id: u32,
    #[serde(default, skip_serializing_if = "is_false")]
    pub steam_restarted: bool,
}

/// Shortcut deletion with restart option.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteShortcutWithRestartRequest {
    pub user_id: String,
    pub app_id: u32,
    #[serde(default, skip_serializing_if = "is_false")]
    pub restart_steam: bool,
}

/// Result of game deletion.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteGameResponse {
    pub status: String,
    pub game_name: String,
    pub steam_restarted: bool,
}

// ---------------------------------------------------------------------------
// Artwork payloads
// ---------------------------------------------------------------------------

/// Artwork operation result.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ArtworkResponse {
    pub applied: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub failed: Vec<ArtworkFailed>,
}

/// A failed artwork application.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ArtworkFailed {
    #[serde(rename = "type")]
    pub art_type: String,
    pub error: String,
}

/// Binary artwork image transfer acknowledgment.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ArtworkImageResponse {
    pub success: bool,
    pub artwork_type: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub error: String,
}

// ---------------------------------------------------------------------------
// Operation payloads
// ---------------------------------------------------------------------------

/// Generic result for operations.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct OperationResult {
    pub success: bool,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub message: String,
}

/// Push notification for operation progress.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OperationEvent {
    #[serde(rename = "type")]
    pub event_type: String,
    pub status: String,
    pub game_name: String,
    pub progress: f64,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub message: String,
}

// ---------------------------------------------------------------------------
// Steam control payloads
// ---------------------------------------------------------------------------

/// Result of Steam restart.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RestartSteamResponse {
    pub success: bool,
    pub message: String,
}

// ---------------------------------------------------------------------------
// Upload extended payloads
// ---------------------------------------------------------------------------

/// Upload request with file manifest.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InitUploadRequestFull {
    pub config: UploadConfig,
    pub total_size: i64,
    pub files: Vec<FileEntry>,
}

/// A file in the upload manifest.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FileEntry {
    pub relative_path: String,
    pub size: i64,
}

/// Upload response with chunk size configuration.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InitUploadResponseFull {
    pub upload_id: String,
    pub chunk_size: i32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub resume_from: Option<HashMap<String, i64>>,
}

/// Upload chunk with full metadata.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UploadChunkRequestFull {
    pub upload_id: String,
    pub file_path: String,
    pub offset: i64,
    pub size: i32,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub checksum: String,
}

/// Upload completion with shortcut configuration.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompleteUploadRequestFull {
    pub upload_id: String,
    pub create_shortcut: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub shortcut: Option<ShortcutConfig>,
}

/// Upload completion result with path and app ID.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompleteUploadResponseFull {
    pub success: bool,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub path: String,
    #[serde(default, skip_serializing_if = "is_zero_u32")]
    pub app_id: u32,
}

/// Upload progress event.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UploadProgressEvent {
    pub upload_id: String,
    pub transferred_bytes: i64,
    pub total_bytes: i64,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub current_file: String,
    pub percentage: f64,
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn is_zero_i64(v: &i64) -> bool {
    *v == 0
}

fn is_zero_u32(v: &u32) -> bool {
    *v == 0
}

fn is_false(v: &bool) -> bool {
    !v
}

/// Custom base64 serde module matching Go's automatic `[]byte` ↔ base64 encoding.
mod base64_bytes {
    use base64::{Engine, engine::general_purpose::STANDARD};
    use serde::{Deserialize, Deserializer, Serialize, Serializer};

    pub fn serialize<S: Serializer>(data: &[u8], serializer: S) -> Result<S::Ok, S::Error> {
        STANDARD.encode(data).serialize(serializer)
    }

    pub fn deserialize<'de, D: Deserializer<'de>>(deserializer: D) -> Result<Vec<u8>, D::Error> {
        let s = String::deserialize(deserializer)?;
        STANDARD.decode(&s).map_err(serde::de::Error::custom)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn upload_chunk_base64_roundtrip() {
        let req = UploadChunkRequest {
            upload_id: "u1".into(),
            offset: 0,
            data: vec![0x48, 0x65, 0x6c, 0x6c, 0x6f],
            file_path: "test.bin".into(),
            is_last: false,
        };
        let json = serde_json::to_string(&req).unwrap();
        // Go encodes []byte as base64, "Hello" = "SGVsbG8="
        assert!(json.contains("SGVsbG8="));
        let parsed: UploadChunkRequest = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed.data, vec![0x48, 0x65, 0x6c, 0x6c, 0x6f]);
    }

    #[test]
    fn hub_connected_omit_empty() {
        let req = HubConnectedRequest {
            name: "Hub".into(),
            version: "0.1.0".into(),
            platform: String::new(),
            hub_id: String::new(),
            token: String::new(),
            protocol_version: 0,
        };
        let json = serde_json::to_string(&req).unwrap();
        assert!(!json.contains("platform"));
        assert!(!json.contains("hubId"));
        assert!(!json.contains("token"));
        assert!(!json.contains("protocolVersion"));
    }

    #[test]
    fn hub_connected_protocol_version_present() {
        let req = HubConnectedRequest {
            name: "Hub".into(),
            version: "0.1.0".into(),
            platform: String::new(),
            hub_id: String::new(),
            token: String::new(),
            protocol_version: 1,
        };
        let json = serde_json::to_string(&req).unwrap();
        assert!(json.contains("\"protocolVersion\":1"));
    }

    #[test]
    fn hub_connected_legacy_json_defaults_to_zero() {
        let json = r#"{"name":"Hub","version":"0.1.0"}"#;
        let req: HubConnectedRequest = serde_json::from_str(json).unwrap();
        assert_eq!(req.protocol_version, 0);
    }

    #[test]
    fn agent_status_response_roundtrip() {
        let resp = AgentStatusResponse {
            name: "Agent".into(),
            version: "0.6.0".into(),
            platform: "steamdeck".into(),
            accept_connections: true,
            telemetry_enabled: true,
            telemetry_interval: 5,
            console_log_enabled: false,
            protocol_version: 1,
        };
        let json = serde_json::to_string(&resp).unwrap();
        assert!(json.contains("\"acceptConnections\":true"));
        assert!(json.contains("\"protocolVersion\":1"));
        let parsed: AgentStatusResponse = serde_json::from_str(&json).unwrap();
        assert_eq!(resp, parsed);
    }

    #[test]
    fn agent_status_legacy_json_defaults_to_zero() {
        let json = r#"{
            "name":"Agent","version":"0.6.0","platform":"steamdeck",
            "acceptConnections":true,"telemetryEnabled":true,
            "telemetryInterval":5,"consoleLogEnabled":false
        }"#;
        let resp: AgentStatusResponse = serde_json::from_str(json).unwrap();
        assert_eq!(resp.protocol_version, 0);
    }

    #[test]
    fn init_upload_response_full_roundtrip() {
        let mut resume = HashMap::new();
        resume.insert("file1.bin".to_string(), 1024i64);
        let resp = InitUploadResponseFull {
            upload_id: "u1".into(),
            chunk_size: 1048576,
            resume_from: Some(resume),
        };
        let json = serde_json::to_string(&resp).unwrap();
        let parsed: InitUploadResponseFull = serde_json::from_str(&json).unwrap();
        assert_eq!(resp, parsed);
    }

    #[test]
    fn operation_event_roundtrip() {
        let evt = OperationEvent {
            event_type: "install".into(),
            status: "progress".into(),
            game_name: "TestGame".into(),
            progress: 42.5,
            message: String::new(),
        };
        let json = serde_json::to_string(&evt).unwrap();
        // Go uses `json:"type"`, Rust uses `#[serde(rename = "type")]`
        assert!(json.contains("\"type\":\"install\""));
        assert!(!json.contains("message"));
        let parsed: OperationEvent = serde_json::from_str(&json).unwrap();
        assert_eq!(evt, parsed);
    }

    #[test]
    fn delete_shortcut_omit_empty() {
        let req = DeleteShortcutRequest {
            user_id: 12345,
            app_id: 0,
            name: String::new(),
        };
        let json = serde_json::to_string(&req).unwrap();
        assert!(!json.contains("appId"));
        assert!(!json.contains("name"));
    }

    #[test]
    fn artwork_failed_type_field() {
        let f = ArtworkFailed {
            art_type: "hero".into(),
            error: "file not found".into(),
        };
        let json = serde_json::to_string(&f).unwrap();
        assert!(json.contains("\"type\":\"hero\""));
    }

    #[test]
    fn game_log_wrapper_status_map_keys() {
        // Go serializes map[uint32]bool with string keys
        let json = r#"{"wrappers":{"12345":true,"67890":false}}"#;
        let parsed: super::super::telemetry::GameLogWrapperStatusEvent =
            serde_json::from_str(json).unwrap();
        assert_eq!(parsed.wrappers.get(&12345), Some(&true));
        assert_eq!(parsed.wrappers.get(&67890), Some(&false));
    }
}
